source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -55,14 +56,27 @@ class CertificateProvider {
   virtual const std::string& getCACertificate(absl::string_view cert_name) const PURE;
 
   /*
-   * Get TLS identity certificates directly from provider
+   * Get TLS certpair directly from provider, return identity cert and key for handshake
+   * or ca cert and key for issuer.
    */
-  virtual std::list<Certpair> getIdentityCertificates(absl::string_view cert_name) PURE;
+  virtual std::list<Certpair> getCertpair(absl::string_view cert_name) PURE;
 
   /*
    * Generate TLS identity certificate dynamically during TLS handshake
    */
-  virtual Certpair* generateTlsCertificate(absl::string_view server_name) PURE;
+  virtual Certpair* generateIdentityCertificate(const SSL_CLIENT_HELLO* ssl_client_hello) PURE;
+
+  /*
+   * CertificateSubsriptionCallbacks
```

Comment:
Why are these callbacks part of the `CertificateProvider` API?  It seems like these would need to be invoked on all registered subscribers by the provider impl, which doesn't require them to be methods of the provider API itself.

Reply:
CertificateProvider inherit from CertificateSubscriptionCallbacks (I forget updating that), the CertificateProvider impl* is going to be passed to subscription constructor, when subscription watches updates it invokes corresponding callbacks.

Reply:
You might want to follow pattern like RDS or SDS where multiple subscription is in use. The callbacks shouldn't be part of the interface, but your impl can implement both if needed.

Reply:
@markdroth @lizan please have a look at updated patch.

I think I misunderstood something about asynchronous mechanism before. Currently there is one addUpdateCallback for caller to register callback functions, certificate provider will invoke this callback functions when a cert change occurs.

Besides, I remove generateIdentityCertificate interface temporarily, because I don't find a proper way to support that, still investigating that.